cardano-cli/src/Cardano/CLI/Compatible/Exception.hs

```diff
@@ -0,0 +1,49 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Cardano.CLI.Compatible.Exception
+  ( CustomCliException (..)
+  , throwCliError
+  , fromEitherCli
+  , fromEitherIOCli
+  )
+where
+
+import           Cardano.Api
+
+import           GHC.Stack
+
+import           RIO
+
+-- | Custom exception type for CLI commands. Any custom errors created
+-- in `cardano-cl` should be wrapped in this exception type.
+data CustomCliException where
+  CustomCliException
+    :: (Show error, Typeable error, Error error)
+    => error -> CallStack -> CustomCliException
+
+deriving instance Show CustomCliException
+
+instance Exception CustomCliException where
+  displayException (CustomCliException e cStack) =
+    unlines
+      [ show (prettyError e)
+      , prettyCallStack cStack
+      ]
+
+throwCliError :: MonadIO m => CustomCliException -> m a
+throwCliError = throwIO
+
+fromEitherCli :: (HasCallStack, MonadIO m, Show e, Typeable e, Error e) => Either e a -> m a
+fromEitherCli = \case
+  Left e -> throwCliError $ CustomCliException e callStack
+  Right a -> return a
+
+fromEitherIOCli :: (HasCallStack, MonadIO m, Show e, Typeable e, Error e) => IO (Either e a) -> m a
+fromEitherIOCli action = do
+  result <- liftIO action
```

Comment:
This is more complicated than just `liftIO`. We need to catch all synchronous exceptions here and wrap them, and also provide the call stack. So basically we need analogous wrapper to `CustomCliException` but for `Exception`s this time.

Reply:
We could make this even more robust by making CIO a newtype, and not providing `MonadIO`: https://github.com/IntersectMBO/cardano-cli/commit/3c1f6f038d1f3f6cc4d4c5077a7b9dc2791f48ad

Reply:
Can you update the ADR with this first?